torch_np/_ndarray.py

```diff
@@ -194,129 +162,81 @@ def __len__(self):
     ### arithmetic ###
 
     # add, self + other
-    def __add__(self, other):
-        return _ufunc_impl.add(self, asarray(other))
-
-    def __radd__(self, other):
-        return _ufunc_impl.add(self, asarray(other))
+    __add__ = __radd__ = _binary_ufuncs.add
 
     def __iadd__(self, other):
-        return _ufunc_impl.add(self, asarray(other), out=self)
+        return _binary_ufuncs.add(self, other, out=self)
 
 
     # sub, self - other
-    def __sub__(self, other):
-        return _ufunc_impl.subtract(self, asarray(other))
-
-    def __rsub__(self, other):
-        return _ufunc_impl.subtract(self, asarray(other))
+    __sub__ = __rsub__ = _binary_ufuncs.subtract
 
     def __isub__(self, other):
-        return _ufunc_impl.subtract(self, asarray(other), out=self)
+        return _binary_ufuncs.subtract(self, other, out=self)
 
 
     # mul, self * other
-    def __mul__(self, other):
-        return _ufunc_impl.multiply(self, asarray(other))
-
-    def __rmul__(self, other):
-        return _ufunc_impl.multiply(self, asarray(other))
+    __mul__ = __rmul__ = _binary_ufuncs.multiply
 
     def __imul__(self, other):
-        return _ufunc_impl.multiply(self, asarray(other), out=self)
+        return _binary_ufuncs.multiply(self, other, out=self)
 
 
     # div, self / other
-    def __truediv__(self, other):
-        return _ufunc_impl.divide(self, asarray(other))
-
-    def __rtruediv__(self, other):
-        return _ufunc_impl.divide(self, asarray(other))
+    __truediv__ = __rtruediv__ = _binary_ufuncs.divide
 
     def __itruediv__(self, other):
-        return _ufunc_impl.divide(self, asarray(other), out=self)
+        return _binary_ufuncs.divide(self, other, out=self)
 
 
     # floordiv, self // other
-    def __floordiv__(self, other):
-        return _ufunc_impl.floor_divide(self, asarray(other))
-
-    def __rfloordiv__(self, other):
-        return _ufunc_impl.floor_divide(self, asarray(other))
+    __floordiv__ = __rfloordiv__ = _binary_ufuncs.floor_divide
 
     def __ifloordiv__(self, other):
-        return _ufunc_impl.floor_divide(self, asarray(other), out=self)
+        return _binary_ufuncs.floor_divide(self, other, out=self)
```

Comment:
Nit, these can be done with a partial (+ a wraps if you want to preserve the name)

Reply:
Not easily because it needs to bind `self`:

```diff
diff --git a/torch_np/_ndarray.py b/torch_np/_ndarray.py
index e01fd92..2b1adf4 100644
--- a/torch_np/_ndarray.py
+++ b/torch_np/_ndarray.py
@@ -163,10 +163,7 @@ class ndarray:
 
     # add, self + other
     __add__ = __radd__ = _binary_ufuncs.add
-
-    def __iadd__(self, other):
-        return _binary_ufuncs.add(self, other, out=self)
-
+    __iadd__ = functools.partialmethod(_binary_ufuncs.add, out=self)
 
     # sub, self - other
     __sub__ = __rsub__ = _binary_ufuncs.subtract
```

triggers

```
$ pytest torch_np/ -x

<snip>

    class ndarray:
torch_np/_ndarray.py:166: in ndarray
    __iadd__ = functools.partialmethod(_binary_ufuncs.add, out=self)
E   NameError: name 'self' is not defined

```

Reply:
You are completely right, nvm.